Add plan/class filter options to svcat get instances

artmello opened the pull request:
Closes: #1727

luksa commented on cmd/svcat/command/command.go:
```diff
@@ -144,6 +164,48 @@ func determineOutputFormat(flags *pflag.FlagSet) (string, error) {
+func determinePlanFilter(flags *pflag.FlagSet) string {
+	plan, _ := flags.GetString("plan")
+
+	if plan != "" {
```
This if statement and the code that follows can be simplified to:
```
return plan
```

luksa commented on cmd/svcat/command/command.go:
```diff
@@ -144,6 +164,48 @@ func determineOutputFormat(flags *pflag.FlagSet) (string, error) {
+func determineClassFilter(flags *pflag.FlagSet) string {
+	class, _ := flags.GetString("class")
+
+	if class != "" {
```
This can also be simplified to `return class`.

luksa commented on cmd/svcat/instance/get_cmd.go:
```diff
@@ -75,6 +90,24 @@ func (c *getCmd) getAll() error {
+		instances = c.filterListByPlan(instances)
+
+		if len(instances.Items) <= 0 {
+			// All instances found were filtered out by plan
```
I personally don't think this comment is necessary, as the code seems clear enough.

luksa commented on cmd/svcat/instance/get_cmd.go:
```diff
@@ -75,6 +90,24 @@ func (c *getCmd) getAll() error {
+		instances = c.filterListByClass(instances)
+
+		if len(instances.Items) <= 0 {
+			// All instances found were filtered out by class
```
Same here.

luksa commented on cmd/svcat/instance/get_cmd.go:
```diff
@@ -85,7 +118,61 @@ func (c *getCmd) get() error {
+	return &p
+}
+
+func (c *getCmd) filterByPlan(instance *v1beta1.ServiceInstance) bool {
```
The name of this method isn't ideal. It's not clear whether it returns `true` or `false` when the `instance`'s plan matches the filter. Can you rename this to something like `acceptedByPlanFilter` or `planFilterAccepts`?

luksa commented on cmd/svcat/instance/get_cmd.go:
```diff
@@ -85,7 +118,61 @@ func (c *getCmd) get() error {
+	return &p
+}
+
+func (c *getCmd) filterByClass(instance *v1beta1.ServiceInstance) bool {
```
Same as above (naming).

luksa commented on cmd/svcat/instance/get_cmd.go:
```diff
@@ -85,7 +118,61 @@ func (c *getCmd) get() error {
+			continue
+		}
+
+		p.Items = append(p.Items, instance)
```
This for loop could be simplified to:
```
if c.filterByPlan(&instance) {
    p.Items = append(p.Items, instance)
}
```

luksa commented:
Also, please add test cases to `TestCommandOutput()` in `cmd/svcat/svcat_test.g`

artmello commented:
@luksa Thanks for reviewing this. I think latest commits should address all your comments, specially about different output between filtering out all entries and getting all entries when there is no instances at all. About code improvements, most of that code was removed after latest changes, but, if you have some time, I would really appreciate if you let me know of any other suggestion for improvements. I also added new tests as suggested.

@carolynvs I moved the --plan/--class flags out from command.go and created PlanFilteredCommand and ClassFilteredCommand, following WaitableCommand as example. To be honest, I am not 100% sure about my division. I think it could be either a FilteredCommand (grouping all possible filter flags) or a, more generic, PlanSpecifiedCommand (to be used by any command that has the --plan flag, with no concerns about the final use of such flag). Let me know if you agree with these commands or if you think some other arrangement would be better. Following your request I also moved the filter logic out of cmd/svcat to pkg/svcat/servicecatalog. But, trying to be consistent I changed both RetrieveInstances and RetrieveInstance, so all filtered logic is there by now. One problem is that RetrieveInstace is used on different places of the code and I needed to add two empty parameters to those calls. Not sure if this is the best approach. Once again, thx for reviewing.

artmello commented:
@luksa Repositioned parameters to "class" be displayed before "plan" in all places they appear together. Also followed your suggestion about RetrieveInstance error messages when instance is found but filtered out by class/plan.
I agree with your point about not using name and class/plan filter at the same time to retrieve an instance. Will wait @carolynvs opinion and update the PR with decided approach, if necessary

carolynvs commented on pkg/svcat/service-catalog/instance.go:
```diff
@@ -34,21 +34,50 @@ const (
 
 // RetrieveInstance gets an instance by its name.
-func (sdk *SDK) RetrieveInstance(ns, name string) (*v1beta1.ServiceInstance, error) {
+func (sdk *SDK) RetrieveInstance(ns, name, classFilter, planFilter string) (*v1beta1.ServiceInstance, error) {
```
This function shouldn't have been modified to accept a class and plan filter. It retrieves a single instance by name, so filtering doens't really apply here.

How about we add check to the `Validate` method on the commands that checks for the name and the class/plan filters being set at the same time.

carolynvs commented on cmd/svcat/command/classfiltered.go:
```diff
@@ -0,0 +1,57 @@
+)
+
+// HasClassFlag represents a command that supports --class.
+type HasClassFlag interface {
```
👍 Thanks this is just what I was hoping for!

carolynvs commented on cmd/svcat/command/classfiltered.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+// ClassFilteredCommand adds support to a command for the --class flag.
+type ClassFilteredCommand struct {
```
FYI, Morgan and I decided on a slightly different naming convention so that it's not so stuttered/verbose

https://github.com/kubernetes-incubator/service-catalog/blob/0eb083fd800e3d0bc658e5ccc285951c2c3e33a1/cmd/svcat/instance/get_cmd.go#L37

Basically since the package is already called `command`, we don't need the `*Command` suffix on the type and the constructor below.

I'm cool with that being in a follow-up issue/PR though. I realize we changed that out from under you and I apologize for the churn! 😊

artmello commented:
@carolynvs latest commits should hopefully address your comments on this. And since we are not filtering instances get by name any more, I think it should also address the pending one from @luksa. Thanks for the suggestions!

artmello commented:
@carolynvs latest commits should hopefully address your comments on this. And since we are not filtering instances get by name any more, I think it should also address the pending one from @luksa. Thanks for the suggestions!

luksa commented:
LGTM, but I don't have the required privileges to add the LGTM2 label.

carolynvs commented:
@luksa I am working with Paul and co to get more of y'all reviewer access. Hopefully within the next couple days! 👍

carolynvs commented:
I added the LGTM2 for ya 😉
